File: backend/src/main/java/kr/co/softice/mes/domain/service/RoleService.java
package kr.co.softice.mes.domain.service;

import kr.co.softice.mes.domain.entity.PermissionEntity;
import kr.co.softice.mes.domain.entity.RoleEntity;
import kr.co.softice.mes.domain.entity.RolePermissionEntity;
import kr.co.softice.mes.domain.entity.TenantEntity;
import kr.co.softice.mes.domain.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * Role Service
 * 역할 관리 서비스
 *
 * @author Moon Myung-seop
 */
@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class RoleService {

    private final RoleRepository roleRepository;
    private final TenantRepository tenantRepository;
    private final PermissionRepository permissionRepository;
    private final RolePermissionRepository rolePermissionRepository;

    /**
     * Find role by ID
     */
    public Optional<RoleEntity> findById(Long roleId) {
        log.debug("Finding role by ID: {}", roleId);
        return roleRepository.findById(roleId);
    }

    /**
     * Find role by tenant and role code
     */
    public Optional<RoleEntity> findByTenantAndRoleCode(String tenantId, String roleCode) {
        log.debug("Finding role by tenant: {} and code: {}", tenantId, roleCode);
        return roleRepository.findByTenant_TenantIdAndRoleCode(tenantId, roleCode);
    }

    /**
     * Find roles by tenant
     */
    public List<RoleEntity> findByTenant(String tenantId) {
        log.debug("Finding roles by tenant: {}", tenantId);
        return roleRepository.findByTenantIdWithTenant(tenantId);
    }

    /**
     * Find active roles by tenant
     */
    public List<RoleEntity> findActiveRolesByTenant(String tenantId) {
        log.debug("Finding active roles by tenant: {}", tenantId);
        return roleRepository.findByTenantIdAndIsActiveWithTenant(tenantId, true);
    }

    /**
     * Find permissions by role
     */
    public List<PermissionEntity> findPermissionsByRole(Long roleId) {
        log.debug("Finding permissions for role: {}", roleId);

        RoleEntity role = roleRepository.findById(roleId)
                .orElseThrow(() -> new IllegalArgumentException("Role not found: " + roleId));

        return rolePermissionRepository.findPermissionsByRole(role);
    }

    /**
     * Create new role
     */
    @Transactional
    public RoleEntity createRole(RoleEntity role) {
        log.info("Creating new role: {} for tenant: {}", role.getRoleCode(), role.getTenant().getTenantId());

        // Check if role code already exists for tenant
        if (roleRepository.existsByTenantAndRoleCode(role.getTenant(), role.getRoleCode())) {
            throw new IllegalArgumentException("Role code already exists: " + role.getRoleCode());
        }

        return roleRepository.save(role);
    }

    /**
     * Update role
     */
    @Transactional
    public RoleEntity updateRole(RoleEntity role) {
        log.info("Updating role: {}", role.getRoleId());

        if (!roleRepository.existsById(role.getRoleId())) {
            throw new IllegalArgumentException("Role not found: " + role.getRoleId());
        }

        return roleRepository.save(role);
    }

    /**
     * Delete role
     */
    @Transactional
    public void deleteRole(Long roleId) {
        log.info("Deleting role: {}", roleId);

        RoleEntity role = roleRepository.findById(roleId)
                .orElseThrow(() -> new IllegalArgumentException("Role not found: " + roleId));

        // Delete all role-permission mappings first
        rolePermissionRepository.deleteByRole(role);

        // Delete role
        roleRepository.deleteById(roleId);
    }

    /**
     * Assign permission to role
     */
    @Transactional
    public RolePermissionEntity assignPermission(Long roleId, Long permissionId) {
        log.info("Assigning permission {} to role {}", permissionId, roleId);

        RoleEntity role = roleRepository.findById(roleId)
                .orElseThrow(() -> new IllegalArgumentException("Role not found: " + roleId));

        PermissionEntity permission = permissionRepository.findById(permissionId)
                .orElseThrow(() -> new IllegalArgumentException("Permission not found: " + permissionId));

        // Check if mapping already exists
        if (rolePermissionRepository.existsByRoleAndPermission(role, permission)) {
            throw new IllegalArgumentException("Permission already assigned to role");
        }

        RolePermissionEntity rolePermission = RolePermissionEntity.builder()
                .role(role)
                .permission(permission)
                .build();

        return rolePermissionRepository.save(rolePermission);
    }

    /**
     * Remove permission from role
     */
    @Transactional
    public void removePermission(Long roleId, Long permissionId) {
        log.info("Removing permission {} from role {}", permissionId, roleId);

        RoleEntity role = roleRepository.findById(roleId)
                .orElseThrow(() -> new IllegalArgumentException("Role not found: " + roleId));

        PermissionEntity permission = permissionRepository.findById(permissionId)
                .orElseThrow(() -> new IllegalArgumentException("Permission not found: " + permissionId));

        List<RolePermissionEntity> rolePermissions = rolePermissionRepository.findByRole(role);
        rolePermissions.stream()
                .filter(rp -> rp.getPermission().equals(permission))
                .findFirst()
                .ifPresent(rolePermissionRepository::delete);
    }
}

File: frontend/src/services/roleService.ts
/**
 * Role Service
 * 역할 관리 API 서비스
 * @author Moon Myung-seop
 */

import apiClient from './api';
import {
  Role,
  RoleCreateRequest,
  RoleUpdateRequest,
  Permission,
} from '@/types';

class RoleService {
  private basePath = '/roles';

  /**
   * 역할 목록 조회
   */
  async getRoles(): Promise<Role[]> {
    const response = await apiClient.get<Role[]>(this.basePath);
    return response.data;
  }

  /**
   * 활성 역할 목록 조회
   */
  async getActiveRoles(): Promise<Role[]> {
    const response = await apiClient.get<Role[]>(`${this.basePath}/active`);
    return response.data;
  }

  /**
   * 역할 상세 조회
   */
  async getRole(roleId: number): Promise<Role> {
    const response = await apiClient.get<Role>(`${this.basePath}/${roleId}`);
    return response.data;
  }

  /**
   * 역할 생성
   */
  async createRole(data: RoleCreateRequest): Promise<Role> {
    const response = await apiClient.post<Role>(this.basePath, data);
    return response.data;
  }

  /**
   * 역할 수정
   */
  async updateRole(roleId: number, data: RoleUpdateRequest): Promise<Role> {
    const response = await apiClient.put<Role>(`${this.basePath}/${roleId}`, data);
    return response.data;
  }

  /**
   * 역할 삭제
   */
  async deleteRole(roleId: number): Promise<void> {
    await apiClient.delete(`${this.basePath}/${roleId}`);
  }

  /**
   * 역할의 권한 목록 조회
   */
  async getRolePermissions(roleId: number): Promise<Permission[]> {
    const response = await apiClient.get<Permission[]>(`${this.basePath}/${roleId}/permissions`);
    return response.data;
  }

  /**
   * 역할에 권한 할당
   */
  async assignPermission(roleId: number, permissionId: number): Promise<void> {
    await apiClient.post(`${this.basePath}/${roleId}/permissions`, { permissionId });
  }

  /**
   * 역할에서 권한 제거
   */
  async removePermission(roleId: number, permissionId: number): Promise<void> {
    await apiClient.delete(`${this.basePath}/${roleId}/permissions/${permissionId}`);
  }

  /**
   * 역할 활성화
   */
  async activateRole(roleId: number): Promise<Role> {
    const response = await apiClient.patch<Role>(`${this.basePath}/${roleId}/activate`);
    return response.data;
  }

  /**
   * 역할 비활성화
   */
  async deactivateRole(roleId: number): Promise<Role> {
    const response = await apiClient.patch<Role>(`${this.basePath}/${roleId}/deactivate`);
    return response.data;
  }
}

export const roleService = new RoleService();
export default roleService;
